central/xlsx/device.go

```diff
@@ -0,0 +1,174 @@
+					// if not, insert the mapping object as a new column in the Devices sheet with defaultValue set in each data row
+					if mapping.defaultValue != "" {
+						err = checkMappingObject(xlsFile, devicesSheetName, &colCount, len(rows), mapping.defaultValue, objectField, &header)
+						if err != nil {
+							return fmt.Errorf("failed to check mapping object: %w", err)
+						}
+					}
+				}
+			}
+		}
+		break
+	}
+
+	rows, err = xlsFile.GetRows(devicesSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s after inserting misshing columns: %w", devicesSheetName, err)
+	}
+
+	// parse the device data rows
+	for rowIndex, row := range rows {
+		if rowIndex == 0 {
+			continue
+		}
+
+		convertedDevice := dtos.Device{ProtocolName: protocol}
+		_, err = readStruct(&convertedDevice, protocol, header, row)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal an excel row into Device DTO: %w", err)
+		}
+
+		// validate the device DTO
+		err := common.Validate(convertedDevice)
+		if err != nil {
+			deviceErr := fmt.Errorf("device %s validation error: %v", convertedDevice.Name, err)
+			deviceXlsx.ValidateErrors = append(deviceXlsx.ValidateErrors, deviceErr)
+		} else {
+			deviceXlsx.Devices = append(deviceXlsx.Devices, &convertedDevice)
+		}
+	}
+
+	err = deviceXlsx.convertAutoEvents(xlsFile)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// convertAutoEvents parses the AutoEvents sheet and convert the rows to AutoEvent DTOs
+func (deviceXlsx *deviceXlsx) convertAutoEvents(xlsFile *excelize.File) error {
+	var header []string
+
+	rows, err := xlsFile.GetRows(autoEventsSheetName)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve all rows from %s worksheet: %w", autoEventsSheetName, err)
+	}
+
+	for rowIndex, row := range rows {
+		// parse the header row
+		if rowIndex == 0 {
+			// get the column count of the header row to see if any Object field from MappingTable sheet is not defined
+			colCount := len(row)
+
+			// assign the first row to header
+			header = row
+
+			// AutoEvents sheet should at least define 4 columns in the header row
+			if colCount < 4 {
+				for objectField, mapping := range deviceXlsx.fieldMappings {
+					if !startsWithAutoEvents(mapping.path) {
+						// if the mapping path doesn't start with autoEvents, skip the check of the AutoEvents sheet header column
+						continue
+					}
+
+					// check if the mapping object is defined in the AutoEvents sheet if the defaultValue is not empty
+					// if not, insert the mapping object as a new column in the Devices sheet with defaultValue set in each data row
+					err = checkMappingObject(xlsFile, autoEventsSheetName, &colCount, len(rows), mapping.defaultValue, objectField, &header)
+					if err != nil {
+						return fmt.Errorf("failed to check mapping object: %w", err)
+					}
+				}
+			}
+		}
+		break
+	}
+
+	rows, err = xlsFile.GetRows(autoEventsSheetName)
```

Comment:
why do you need to retrieve auto event rows twice?  can `rows` be reused?

Reply:
As discussed, the total columns might be updated from mappingTable sheet definition so need to call `GetRows` again here.